docs/designs/concepts.md

```diff
@@ -0,0 +1,180 @@
+# Concepts Design Doc
+
+This document describes core concepts of ElasticDL.
+
+An ElasticDL model consists of two kinds of parameters:
+
+1. dense parameters, and
+1. embedding tables.
+
+A dense parameter is a dense tensor with a name.  An embedding table is a map from some ID to embedding vectors.  An embedding table also has a name.  Formalizing the concepts, we have
+
+- model = {dense parameter} + {embedding table}
+- dense parameter = tensor + name
+- embedding table = {id, tensor} + name
+
+where the curly braces denote zero or more.
+
+It is notable that in practice, we use the rule
+
+- embedding table = tensor + index + name
+- index = map[ID]address
+
+because we append all embedding vectors into the tensor and use index, which is a map from ID to the starting address of the embedding vector in the tensor.
+
+To update the model, workers compute and report gradients.  Accordingly, we have two kinds of gradients:
+
+1. dense gradient, and
+1. embedding table gradient
+
+The content of dense gradient is the same as that of the dense parameter.  The content of embedding table gradient is the same as that of the embedding table.
+
+Please be aware that if some embedding table maps zero-based successive ID values to embedding vectors.  Such kind of embedding table, if not too large, could be represented by a dense parameters.  In the rest of the design, the concept embedding table refers to a sparse/general embedding table, where the ID might not be zero-based successive values, and even not an numerical value.
+
+Let's make a short summary, following is all the core concepts of ElasticDL include:
+
+- model = {dense parameter} + {embedding table}
+- dense parameter = tensor + name
+- embedding table = tensor + index + name
+- index = map[ID]address
+
+## In-Memory Representation
+
+EmbeddingTable is initialized lazily in PS. Worker sends item ids to request embedding vectors from PS. If the ID is not found in PS, the corresponding embedding vector will be initialized and sent back to the worker. In order to query and insert embedding vectors efficiently, we use a map data structure to represent EmbeddingTable.
+
+```go
+type TensorDtype = int64
+
+const (
+    Invalid TensorDtype = iota
+    Int8 
+    Int16
+    Int32
+    Int64
+    Float16
+    Float32
+    Float64
+    Bool
+)
+
+type Buffer struct {
```

Comment:
I don't see that we need type Buffer, do we?  If not necessary, let us follow Occam's Razor.

Reply:
Done, remove it